device tree macro: GPIO_DT_SPEC_INST_GET_BY_IDX_OR does not work

**Describe the bug**
Given a binding with the following:
```
    data-bus-gpios:
      type: phandle-array
```
A boards file with:
```
data-bus-gpios =
<0>, <0>, <0>, <0>,
<&gpioa 5 (GPIO_ACTIVE_HIGH)>,
<&gpioa 6 (GPIO_ACTIVE_HIGH)>,
<&gpioa 7 (GPIO_ACTIVE_HIGH)>,
<&gpiod 14 (GPIO_ACTIVE_HIGH)>;
```
And a driver with:
```
                .db_gpios[0] = GPIO_DT_SPEC_INST_GET_BY_IDX_OR(inst, data_bus_gpios, 0, {0}),   \
                .db_gpios[4] = GPIO_DT_SPEC_INST_GET_BY_IDX(inst, data_bus_gpios, 4),           \
```
This fails to compile with the following error:
```
include/zephyr/device.h:83:41: error: '__device_dts_ord_DT_N_S_auxdisplay_P_data_bus_gpios_IDX_0_PH_ORD' undeclared here (not in a function); did you mean 'DT_N_S_auxdisplay_P_data_bus_gpios_IDX_6_PH'?
   83 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~
include/zephyr/sys/util_internal.h:72:26: note: in definition of macro '__DEBRACKET'
   72 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
include/zephyr/sys/util_internal.h:59:9: note: in expansion of macro '__COND_CODE'
   59 |         __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~
include/zephyr/sys/util_macro.h:160:9: note: in expansion of macro 'Z_COND_CODE_1'
  160 |         Z_COND_CODE_1(_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~~~
include/zephyr/devicetree.h:3711:9: note: in expansion of macro 'COND_CODE_1'
 3711 |         COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT),   \
      |         ^~~~~~~~~~~
include/zephyr/sys/util_internal.h:69:53: note: in expansion of macro '__DEBRACKET'
   69 | #define __GET_ARG2_DEBRACKET(ignore_this, val, ...) __DEBRACKET val
      |                                                     ^~~~~~~~~~~
include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
include/zephyr/sys/util_internal.h:59:9: note: in expansion of macro '__COND_CODE'
   59 |         __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~
include/zephyr/sys/util_macro.h:160:9: note: in expansion of macro 'Z_COND_CODE_1'
  160 |         Z_COND_CODE_1(_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~~~
include/zephyr/drivers/gpio.h:339:9: note: in expansion of macro 'COND_CODE_1'
  339 |         COND_CODE_1(DT_NODE_HAS_PROP(node_id, prop),                    \
      |         ^~~~~~~~~~~
include/zephyr/device.h:209:37: note: in expansion of macro 'DEVICE_NAME_GET'
  209 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_ID(node_id))
      |                                     ^~~~~~~~~~~~~~~
include/zephyr/device.h:226:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  226 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))
      |                                  ^~~~~~~~~~~~~~~~~~
include/zephyr/drivers/gpio.h:316:25: note: in expansion of macro 'DEVICE_DT_GET'
  316 |                 .port = DEVICE_DT_GET(DT_GPIO_CTLR_BY_IDX(node_id, prop, idx)),\
      |                         ^~~~~~~~~~~~~
include/zephyr/drivers/gpio.h:340:22: note: in expansion of macro 'GPIO_DT_SPEC_GET_BY_IDX'
  340 |                     (GPIO_DT_SPEC_GET_BY_IDX(node_id, prop, idx)),      \
      |                      ^~~~~~~~~~~~~~~~~~~~~~~
include/zephyr/drivers/gpio.h:392:9: note: in expansion of macro 'GPIO_DT_SPEC_GET_BY_IDX_OR'
  392 |         GPIO_DT_SPEC_GET_BY_IDX_OR(DT_DRV_INST(inst), prop, idx, default_value)
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
drivers/auxdisplay/auxdisplay_hitachi_hd44780.c:301:32: note: in expansion of macro 'GPIO_DT_SPEC_INST_GET_BY_IDX_OR'
  301 |                 .db_gpios[0] = GPIO_DT_SPEC_INST_GET_BY_IDX_OR(inst, data_bus_gpios, 0, {0}),   \
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
samples/drivers/auxdisplay/_BB/zephyr/include/generated/devicetree_generated.h:20865:50: note: in expansion of macro 'HITACHI_HD44780_DEVICE'
20865 | #define DT_FOREACH_OKAY_INST_hitachi_hd44780(fn) fn(0)
      |                                                  ^~
include/zephyr/sys/util_internal.h:105:36: note: in expansion of macro 'DT_FOREACH_OKAY_INST_hitachi_hd44780'
  105 | #define UTIL_PRIMITIVE_CAT(a, ...) a##__VA_ARGS__
      |                                    ^
drivers/auxdisplay/auxdisplay_hitachi_hd44780.c:323:1: note: in expansion of macro 'DT_INST_FOREACH_STATUS_OKAY'
  323 | DT_INST_FOREACH_STATUS_OKAY(HITACHI_HD44780_DEVICE)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
samples/drivers/auxdisplay/_BB/zephyr/include/generated/devicetree_generated.h:1374:37: error: 'DT_N_S_auxdisplay_P_data_bus_gpios_IDX_0_VAL_pin' undeclared here (not in a function); did you mean 'DT_N_S_auxdisplay_P_data_bus_gpios_IDX_5_VAL_pin'?
 1374 | #define DT_N_INST_0_hitachi_hd44780 DT_N_S_auxdisplay
      |                                     ^~~~~~~~~~~~~~~~~
include/zephyr/sys/util_internal.h:72:26: note: in definition of macro '__DEBRACKET'
   72 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
include/zephyr/sys/util_internal.h:59:9: note: in expansion of macro '__COND_CODE'
   59 |         __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~
include/zephyr/sys/util_macro.h:160:9: note: in expansion of macro 'Z_COND_CODE_1'
  160 |         Z_COND_CODE_1(_flag, _if_1_code, _else_code)
```
If the underlying phandle for that index is valid then there is no error, but it seems it cannot cope with a "hole" entry of `<0>`.

**To Reproduce**
As above

**Expected behavior**
Macro to work if it is defined or not

**Impact**
Showstopper for this driver

**Environment (please complete the following information):**
 - OS: Linux
 - Commit SHA or Version used: 8c8d41c32fb5d7a6a965c4cbdfa955d95905cf4a

--- a/include/zephyr/drivers/gpio.h
+++ b/include/zephyr/drivers/gpio.h
@@ -388,8 +388,10 @@ struct gpio_dt_spec {
  * @return static initializer for a struct gpio_dt_spec for the property
  * @see GPIO_DT_SPEC_GET_BY_IDX()
  */
-#define GPIO_DT_SPEC_INST_GET_BY_IDX_OR(inst, prop, idx, default_value)	\
-	GPIO_DT_SPEC_GET_BY_IDX_OR(DT_DRV_INST(inst), prop, idx, default_value)
+#define GPIO_DT_SPEC_INST_GET_BY_IDX_OR(inst, prop, idx, default_value)		\
+	COND_CODE_1(DT_PROP_HAS_IDX(DT_DRV_INST(inst), prop, idx),		\
+		    (GPIO_DT_SPEC_GET_BY_IDX(DT_DRV_INST(inst), prop, idx)),	\
+		    (default_value))
 
 /**
  * @brief Equivalent to GPIO_DT_SPEC_INST_GET_BY_IDX(inst, prop, 0).

